Userland/Libraries/LibDSP/Music.h

```diff
@@ -0,0 +1,153 @@
+/*
+ * Copyright (c) 2021, kleines Filmröllchen <malu.bertsch@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <AK/Types.h>
+#include <AK/Variant.h>
+#include <AK/Vector.h>
+#include <LibAudio/Buffer.h>
+
+namespace LibDSP {
+
+// FIXME: Audio::Frame is 64-bit float, which is quite large for long clips.
+typedef Audio::Frame Sample;
+
+Sample const SAMPLE_OFF = { 0.0, 0.0 };
+
+struct RollNote {
+    u32 length() const { return (off_sample - on_sample) + 1; }
+
+    u32 on_sample;
+    u32 off_sample;
+    u8 pitch;
+    i8 velocity;
+};
+
+enum class SignalType : u8 {
+    Note,
+    Sample
+};
+
+struct Signal : public Variant<Sample, Vector<RollNote>> {
+    SignalType const type;
+
+    Signal()
+        : type(has<Sample>() ? SignalType::Sample : SignalType::Note)
```

Comment:
I'm not too experienced with this, but in this case, wouldn't the `Variant` be empty (because you never initialize it) and thus neither `SignalType` would apply?

Reply:
No, because the initialization of `type` (runtime) happens after the template has been resolved (compile-time). At least it should :^)

Reply:
Okey, it seems I don't understand `Variant` well enough? So I guess it doesn't have a null state.